File: tags/CARDAMOM_V1_1_CE/Tutorial/java/naming_writeExt/src/HelloClient.java
/* =========================================================================== *
 * This file is part of CARDAMOM (R) which is jointly developed by THALES
 * and SELEX-SI.
 * 
 * It is derivative work based on PERCO Copyright (C) THALES 2000-2003.
 * All rights reserved.
 * 
 * CARDAMOM is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Library General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your
 * option) any later version.
 * 
 * CARDAMOM is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Library General Public
 * License for more details.
 * 
 * You should have received a copy of the GNU Library General
 * Public License along with CARDAMOM; see the file COPYING. If not, write to
 * the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 * =========================================================================== */


package cdmw.tutorial;


public class HelloClient {

    private static final int POA_THREAD_POOL_SIZE = 5;
    
    private static final String NAMING_REPOS_URL_OPTION = "--namingRepos-url";
       
    
    private static void usage() {
        System.out.println("Usage:\n");
        System.out.println("helloClient.sh options\n");
        System.out.println("Options:\n");
        System.out.println();
        System.out.println("--help                     Display this information.\n");
        System.out.println();
        System.out.println(NAMING_REPOS_URL_OPTION + "=<URL>  URL of the CDMW Naming and Repository service.\n");
        System.out.println();
    }
    

    
    public static void main(String[] args) {
        
        System.out.println("Start Hello Client");
      
        //
        // help argument
        //
        if(args.length > 0 && args[0].equals("--help"))	{    
            usage();
            return;
        }

        //
        // get Naming & Repository URL from arguments
        //
        String namingReposURL = 
             cdmw.ossupport.OS.getOptionValue (args, NAMING_REPOS_URL_OPTION);
             
        if (namingReposURL.equals("no")) 
        {
            System.out.println("naming and repository url undefined");
            usage();
            return;
        }

        org.omg.CORBA.ORB orb = null;              // orb reference
        
        try  {
            // ===================================================
            // Initialize the ORB
            // ===================================================        
            cdmw.orbsupport.StrategyList orbStrategies = 
                new cdmw.orbsupport.StrategyList();
            orbStrategies.addORBThreaded();
            orbStrategies.addPOAThreadPool(POA_THREAD_POOL_SIZE);
        
            orb = cdmw.orbsupport.ORBSupport.ORBInit(args, orbStrategies);

            // ===================================================
            // Get the root POA 
            // ===================================================
            org.omg.CORBA.Object obj = orb.resolve_initial_references("RootPOA");
            org.omg.PortableServer.POA rootPOA = 
                org.omg.PortableServer.POAHelper.narrow(obj);
            
            // ===================================================
            // Activate the root POA manager
            // ===================================================
            org.omg.PortableServer.POAManager poaManager = rootPOA.the_POAManager();
            poaManager.activate();

            
            // ===================================================
            // Get the naming and repository
            //     from the input URL
            // ===================================================
            com.thalesgroup.CdmwNamingAndRepository.Repository repository = null;
                                      
            org.omg.CORBA.Object obj2 = orb.string_to_object(namingReposURL);
            repository = com.thalesgroup.CdmwNamingAndRepository.RepositoryHelper.narrow(obj2);            
            

            // ======================================================   
            // Initialise the repository interface
            //     "TUTORIAL" has been chosen as the default domain name
            // ======================================================
            cdmw.namingandrepository.RepositoryInterface.init("TUTORIAL", repository);


            // ======================================================
            // Get default root context from repository
            // ======================================================
            org.omg.CosNaming.NamingContext rootContext = 
                repository.resolve_root_context(
                    com.thalesgroup.CdmwNamingAndRepository.DEFAULT_ROOT_CONTEXT.value);            
            

            // ===================================================
            // try to create the "Mike" HelloInterface servant
            // and register it in the repository using cos naming
            // ===================================================
            
            // ===================================================
            // create the "Mike" HelloInterface servant
            // ===================================================
            com.thalesgroup.HelloModule.HelloInterface helloInterfaceMike = null;
            
            {
                HelloInterfaceImpl helloInterfaceServant = new HelloInterfaceImpl("Mike");
    
                // activate servant on POA
                byte[] oid = rootPOA.activate_object(helloInterfaceServant);
                           
                org.omg.CORBA.Object object = rootPOA.id_to_reference(oid);
                helloInterfaceMike = helloInterfaceServant._this(orb);
            }
	    
            // ======================================================
            // Get Naming Interface of Default root context
            // ======================================================
            cdmw.namingandrepository.NamingInterface niDefRoot =
                new cdmw.namingandrepository.NamingInterface(rootContext);
                                                         
             	                
            // =================================================================
            // Try to bind the HelloInterface object in the default root context
            // =================================================================
            try
            {
                niDefRoot.bind ("HelloInterfaceMike", helloInterfaceMike, true);
                System.out.println(">>>> successful binding of HelloInterface for Mike \n" +
                                   "     in the default root context");
            }
            catch(org.omg.CORBA.NO_PERMISSION e)
            {
                System.out.println(">>>> Error in creation of HelloInterface for Mike \n" +
                                   "     in the default root context");
            }           
            
            
                        
            // ===================================================
            // wait for CORBA message
            // ===================================================
            orb.run();
    
        } catch  (cdmw.common.Exception ex) {
            System.err.println(ex.what());
            ex.printStackTrace();
            
        } catch (org.omg.CORBA.SystemException ex) {
            System.err.println(ex.toString());
            ex.printStackTrace();

        } catch (org.omg.CORBA.UserException ex) {
            System.err.println(ex.toString());
            ex.printStackTrace();
        }

        
        // ========================================================
        // program stopping
        // ========================================================

        // ================================================================
        // Release all static object references held by RepositoryInterface
        // ================================================================
        cdmw.namingandrepository.RepositoryInterface.finish();
    
        // =====================================================
        // Destroy orb
        // =====================================================
        if (orb != null) {
            try {
                orb.destroy();
            } catch (org.omg.CORBA.SystemException ex) {
                System.out.println(ex.toString());
                ex.printStackTrace();
            }
        }
    
    }
    
}

File: tags/start/Tutorial/c++/lb_initialization/src/HelloClient.cpp
/* ========================================================================== *
 * This file is part of CARDAMOM (R) which is jointly developed by THALES
 * and SELEX-SI. All rights reserved.
 *
 * CARDAMOM is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Library General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your
 * option) any later version.
 *
 * CARDAMOM is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Library General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Library General
 * Public License along with CARDAMOM; see the file COPYING. If not, write to
 * the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 * ========================================================================= */
                                                                                                 
/**
 * @brief implementation for TestLBInit with cppUnit library.
 *
 * @author Raffaella Panella <rpanella@amjsv.it>
 *
 */

// Standard Files
#include <string>

// Cdmw Files
#include <Foundation/common/Options.hpp>
#include <Foundation/common/String.hpp>
#include <Foundation/osthreads/Thread.hpp>
#include <Foundation/ossupport/OS.hpp>
#include <Foundation/orbsupport/CORBA.hpp>
#include <Foundation/orbsupport/OrbSupport.hpp>
#include <Foundation/orbsupport/StrategyList.hpp>
#include <LoadBalancing/lbinit/LBServiceInit.hpp>
#include "TestUtils.hpp"
#include <LoadBalancing/lbcommon/TAO_LB_IOGRFactory.hpp>
#include <LoadBalancing/idllib/PortableGroup.stub.hpp>
#include "UserDefinedStrategyFactory_impl.hpp"
#include <iostream>
using namespace std;

using namespace Cdmw;


#define ECHO_HEADER() \
    "[Debug] "

#define ECHO_ERROR(comment) \
do {\
    std::cerr << ECHO_HEADER() << " --> " << comment << std::endl;\
} while(0)
#define ENABLE_LB_DEBUG_DUMP 

#ifndef ENABLE_LB_DEBUG_DUMP

#   define DEBUG_DUMP(comment)
#   define DEBUG_ECHO(comment)

#else

#   define DEBUG_DUMP(comment) \
do {\
    std::cerr << ECHO_HEADER() << " --> " << comment << std::endl;\
} while(0)

#   define DEBUG_ECHO(comment) \
do {\
    std::cerr << comment ;\
} while(0)

#endif

int main( int argc, char* argv[] )
{
    try
    {
        char** ORB_init_argv = 0;
        int ORB_init_argc = argc;
        std::string str_opt( "" );

        for (int i=0; i<argc; i++)
            str_opt += std::string( argv[i] ) + " ";
        str_opt += std::string( "--LBInterceptorSide=Client");
        ORB_init_argc += 1;
        ORB_init_argv = Cdmw::Common::String::to_char_array(
            Cdmw::Common::String::to_strings( str_opt ) );

	CORBA::Boolean override_default = false;
	CORBA::Boolean override_force = false;
	std::string policy_override = Cdmw::OsSupport::OS::get_option_value(argc, argv, "--CdmwLBPolicyDefault");
	if (policy_override.compare("yes") && policy_override.compare("no") )
	  override_default = true;
	else
	  {
	    policy_override = Cdmw::OsSupport::OS::get_option_value(argc, argv, "--CdmwLBPolicyForce");
	    if(policy_override.compare("yes") && policy_override.compare("no"))
	      override_force = true;
	  }
        
	if(override_default)
	   DEBUG_DUMP("   Start LB Service overriding IOGR LBPolicy with default value...    ");
	else
	  if(override_force)
	   DEBUG_DUMP("   Start LB Service overriding LBPolicy for all IOGRs   ");
	  else
	    DEBUG_DUMP("   Start LB Service without override IOGR LB Policy   ");
        Cdmw::LB::LBServiceInit::Init(ORB_init_argc, ORB_init_argv);
        
        // Initialises the ORB
        Cdmw::OrbSupport::StrategyList strategyList;
        strategyList.add_OrbThreaded();
        strategyList.add_PoaThreadPerConnection();


        CORBA::ORB_var orb = Cdmw::OrbSupport::OrbSupport::ORB_init(argc, argv, strategyList);

        DEBUG_DUMP("   Register a new User Defined Strategy (A Round Robin, but the next replica must be selected after two requests)   ");
        Cdmw::LB::IOGRFactory* iogr_factory = new Cdmw::LB::TAO_LB_IOGRFactory(std::cerr, orb.in());
        CdmwLB::StrategyInfo* stgy_info = new CdmwLB::StrategyInfo();
        stgy_info->strategy_name = "UserDefinedStrategy";
        stgy_info->strategy_factory = new Cdmw::UserDefinedStrategyFactory_impl(iogr_factory);;
        Cdmw::LB::LBServiceInit::Register_lb_policy(stgy_info);
    

        
        CdmwLBInit::HelloInterface_var hello_group = Cdmw::LB::TestUtils::Get_hello_ref_from_file(orb.in(), "hello_group_default");

        DEBUG_DUMP("   Call hello on Hello Object Group with Default=RoundRobin policy    ");
	for(int i = 0; i<16; i++)
	  {
	    DEBUG_DUMP("  #####HELLO##### from " <<  CORBA::string_dup(hello_group->hello()));
	    Cdmw::OsSupport::OS::sleep(800);	    
            }

       	hello_group = Cdmw::LB::TestUtils::Get_hello_ref_from_file(orb.in(), "hello_group_random");

        DEBUG_DUMP("   Call hello on Hello Object Group with Random policy   ");
	for(int i = 0; i<16; i++)
	  {
	    DEBUG_DUMP("  #####HELLO##### from " <<  CORBA::string_dup(hello_group->hello()));
	    Cdmw::OsSupport::OS::sleep(800);
	  }

        DEBUG_DUMP("   Try to call hello on Hello Object on host 1(The request must be filtered through the client interceptor)   ");
        CdmwLBInit::HelloInterface_var hello_object = Cdmw::LB::TestUtils::Get_hello_ref_from_file(orb.in(), "hello1");
	DEBUG_DUMP("  #####HELLO##### from " <<  CORBA::string_dup(hello_group->hello()));


	

	std::cerr << "Hit a key and ENTER...." << std::endl;
	std::string dummy;
	cin >> dummy;
        }
    catch( const CORBA::Exception &e )
    {
        cerr <<__FILE__<<" "<<__LINE__<<" "<< e._name() << endl;
        return -1;
    }
    catch( const Cdmw::Exception &e )
    {
        cerr <<__FILE__<<" "<<__LINE__<<" "<< e.what() << endl;
        return -1;
    }
    catch(...)
    {
        cerr << "Unexpected exception" << endl;
        return -1;
    }
}
    
